### PR TITLE
fix(compiler): resources cannot reference free variables inflight

### LINKED ISSUE #1447: Resource Declaration: generate implicit constructors based on captures from outer scope
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### Feature Spec

Consider this use case where `global_bucket` is defined outside of the scope of `MyResource`, and is captured by one of `MyResource`'s inflight methods. We would like our JSified code to initialize MyResource and MyResourceClient with this captured global_bucket. 
This is the desired wing code and the js it compiles to:
```ts
// Wing
bring cloud;
​
let global_bucket = new cloud.Bucket();
​
resource MyResource {
  b2: cloud.Bucket;
  init(b2) {
    this.b2 = b2;
  }
​
  inflight my_put() {
    global_bucket.put('key', 'value');
    this.b2.put('key', 'value');
  }
}
​
let someb2 = new cloud.Bucket();
new MyResource(someb2);
​
// JSified preflight
let cloud = require('wingsdk/cloud');
let global_bucket = new cloud.Bucket(this, 'cloud.Bucket');
​
class MyResource {
  constructor(scope, id, b2, captured_global_bucket) {
    super(scope, id);
    this.b2 = b2;
    this.captured_global_bucket = captured_global_bucket;
  }
}
​
new MyResource(this, 'MyResource', b2, global_bucket);
​
// JSified client
class MyResourceClient {
  constructor(b2, captured_global_bucket) {
    this.b2 = b2;
    this.captured_global_bucket = captured_global_bucket;
  }
​
  myPut() {
    this.captured_global_bucket.put('key', 'value');
    this.b2.put('key', 'value');
  }
}
``` 

### Use Cases

See above.

### Implementation Notes

_No response_

### Component

Compiler

### PR DESCRIPTION
**Pull request by Chriscbr**
Closes #1447

This pull request adds support for capturing [free variables](https://en.wikipedia.org/wiki/Free_variables_and_bound_variables) in resource classes, a feature that allows Wing to generate more expressive code for inflight clients. It implements a new visitor, modifies the code generation, and adds a new test case and a snapshot to verify the functionality.

The approach I've taken differs from the solution outlined in #1447. The original issue suggests that we transform free variables into resource fields. I saw some risks with this approach since all usages of the global value inside inflight constructors and inflight methods would also need to be transformed (for example, changing `global_bucket` to `this.global_bucket`), and tracking the list of these mappings so that they only apply to inflight code, and preventing name conflicts, could be tricky.

Instead, this PR augments jsification so that free variables are effectively also emitted as free variables in the JavaScript client code. 

I'm not sure this approach is any better, but it seems to work just as well for a number of cases I've tried.

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/jsify.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use itertools::Itertools;
 use std::{
 	cell::RefCell,
 	cmp::Ordering,
-	collections::{BTreeMap, BTreeSet},
+	collections::{BTreeMap, BTreeSet, HashSet},
```
Comment (Chriscbr) [inline libs/wingc/src/jsify.rs:10]: Fix

File: examples/tests/valid/resource_captures_globals.w [hunk 2]
```diff
@@ -0,0 +1,62 @@
+bring cloud;
```
Comment (eladb) [inline examples/tests/valid/resource_captures_globals.w:1]: Beautiful!

File: libs/wingc/src/ast.rs [hunk 3]
```diff
@@ -23,6 +23,12 @@ impl Symbol {
 			span: Default::default(),
 		}
 	}
+
+	/// Returns true if the symbols refer to the same name AND location in the source code.
+	/// Use `eq` to compare symbols only by name.
+	pub fn exact_eq(&self, other: &Self) -> bool {
```
Comment (eladb) [inline libs/wingc/src/ast.rs:29]: Perhaps call it `same`?

File: libs/wingc/src/jsify.rs [hunk 4]
```diff
@@ -1476,13 +1524,25 @@ impl Display for Component<'_> {
...
-	fn analyze_expr(node: &'a Expr) -> Vec<Component> {
+	fn analyze_expr(&self, node: &'a Expr) -> Vec<Component> {
 		match &node.kind {
 			ExprKind::Reference(Reference::Identifier(x)) => {
+				let scope = match &self.function_def.body {
+					FunctionBody::Statements(scope) => scope,
+					FunctionBody::External(_) => panic!("unexpected expression inside body of extern functions"),
+				};
+				let env = scope.env.borrow();
+				let env = env.as_ref().expect("scope should have an env");
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1535]: Wow this is interesting... overriding `env`... I guess that's legitimate.

File: libs/wingc/src/jsify.rs [hunk 5]
```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
...
+	fn visit_stmt(&mut self, stmt: &Stmt) {
+		match &stmt.kind {
+			// this statement introduces bound variables!
+			StmtKind::Bring {
+				module_name,
+				identifier,
+			} => {
+				self.visit_symbol(&module_name);
+
+				// bring cloud;
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1640]: Nit: Add `"` to indicate this is what we are matching (or some other way)

```suggestion
				// "bring cloud;"
```

File: libs/wingc/src/jsify.rs [hunk 6]
```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
...
 				return [obj, prop].concat();
 			}
 
 			_ => vec![],
 		}
 	}
 }
+
+struct FreeVariableScanner {
+	bound_vars: Vec<Symbol>,
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1598]: Add a wee comment explaining each field

File: libs/wingc/src/jsify.rs [hunk 7]
```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
...
+				// bring "foo" as bar;
+				if let Some(ref id) = identifier {
+					self.visit_symbol(&id);
+
+					// add `bar` to the list of bound variables
+					self.bound_vars.push(id.clone());
+				}
+			}
+			// this statement introduces bound variables!
+			StmtKind::VariableDef {
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1655]:
```suggestion
			// "let x = y"
			StmtKind::VariableDef {
```

File: libs/wingc/src/jsify.rs [hunk 8]
```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
...
+				type_,
+			} => {
+				self.visit_symbol(&var_name);
+				self.visit_expr(&initial_value);
+				if let Some(type_) = type_ {
+					self.visit_type_annotation(&type_);
+				}
+				self.bound_vars.push(var_name.clone());
+			}
+			// invariant: adds net zero bound variables
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1668]: Is this legal in Wing?

```js
for x in y {
  resource Foo {
    init() {
      // access "x"
    }
    
  }
}
```

File: libs/wingc/src/jsify.rs [hunk 9]
```diff
@@ -1520,15 +1580,175 @@ impl<'a> FieldReferenceVisitor<'a> {
...
+			// check if there is already a bound variable with the same name
+			if !self.bound_vars.contains(x) {
+				self.free_vars.push(x.clone());
+			}
+		};
+
+		return visit::visit_reference(self, node);
+	}
+
+	// invariant: adds net zero bound variables
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1623]: Not 100% sure I understand what this means

File: tools/hangar/__snapshots__/test_corpus/resource_captures_globals.w_compile_tf-aws.md [hunk 10]
```diff
@@ -0,0 +1,449 @@
+# [resource_captures_globals.w](../../../../examples/tests/valid/resource_captures_globals.w) | compile | tf-aws
```
Comment (eladb) [inline tools/hangar/__snapshots__/test_corpus/resource_captures_globals.w_compile_tf-aws.md:1]: I just love those markdown snapshots! (@MarkMcCulloh 👍 👍)

**Review by eladb — APPROVED** [review]
Looks great! Some minor comments.

File: libs/wingc/src/jsify.rs (hunk 8 above)
Comment (Chriscbr) [reply libs/wingc/src/jsify.rs:1668]: It probably shouldn't

File: libs/wingc/src/jsify.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/jsify.rs:1535]: Ah yeah, this is a Rust trick. You can also do stuff like this:
```rust
let mut xs = vec![1, 2, 3];
xs.push(4);
let xs = xs; // no longer can be changed!
```
It's pretty neat™️

Comment (eladb) [reply libs/wingc/src/jsify.rs:1535]: Why not. It's safe

**Pull request closed without merging**